org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/WorkspaceSymbolHandler.java

```diff
@@ -82,111 +82,19 @@ public static List<SymbolInformation> search(String query, int maxResults, Strin
 
 			PreferenceManager preferenceManager = JavaLanguageServerPlugin.getPreferencesManager();
 
-			new SearchEngine().searchAllTypeNames(qualifierName == null ? null : qualifierName.toCharArray(), qualifierMatchRule, typeName.toCharArray(), typeMatchRule, IJavaSearchConstants.TYPE, searchScope, new TypeNameMatchRequestor() {
-
-				@Override
-				public void acceptTypeNameMatch(TypeNameMatch match) {
-					try {
-						if (maxResults > 0 && symbols.size() >= maxResults) {
-							return;
-						}
-						Location location = null;
-						try {
-							if (!sourceOnly && match.getType().isBinary()) {
-								location = JDTUtils.toLocation(match.getType().getClassFile());
-							} else if (!match.getType().isBinary()) {
-								location = JDTUtils.toLocation(match.getType());
-							}
-						} catch (Exception e) {
-							JavaLanguageServerPlugin.logException("Unable to determine location for " + match.getSimpleTypeName(), e);
-							return;
-						}
-
-						if (location != null && match.getSimpleTypeName() != null && !match.getSimpleTypeName().isEmpty()) {
-							SymbolInformation symbolInformation = new SymbolInformation();
-							symbolInformation.setContainerName(match.getTypeContainerName());
-							symbolInformation.setName(match.getSimpleTypeName());
-							symbolInformation.setKind(mapKind(match));
-							if (Flags.isDeprecated(match.getType().getFlags())) {
-								if (preferenceManager != null && preferenceManager.getClientPreferences().isSymbolTagSupported()) {
-									symbolInformation.setTags(List.of(SymbolTag.Deprecated));
-								}
-								else {
-									symbolInformation.setDeprecated(true);
-								}
-							}
-							symbolInformation.setLocation(location);
-							symbols.add(symbolInformation);
-							if (maxResults > 0 && symbols.size() >= maxResults) {
-								monitor.setCanceled(true);
-							}
-						}
-					} catch (Exception e) {
-						JavaLanguageServerPlugin.logException("Unable to determine location for " + match.getSimpleTypeName(), e);
-						return;
-					}
-				}
-
-				private SymbolKind mapKind(TypeNameMatch match) {
-					int flags = match.getModifiers();
-					if (Flags.isInterface(flags)) {
-						return SymbolKind.Interface;
-					}
-					if (Flags.isAnnotation(flags)) {
-						return SymbolKind.Property;
-					}
-					if (Flags.isEnum(flags)) {
-						return SymbolKind.Enum;
-					}
-					return SymbolKind.Class;
-				}
-			}, IJavaSearchConstants.WAIT_UNTIL_READY_TO_SEARCH, monitor);
+			SearchEngine engine = new SearchEngine();

```

Comment:
have you considered using a org.eclipse.jdt.core.search.SearchPattern.createOrPattern pattern for both of this type searches and perform a single search operation ?

Reply:
Yes, I thought about this. The part that made me reconsider is https://github.com/eclipse-jdt/eclipse.jdt.core/blob/master/org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/BasicSearchEngine.java#L1787 . There's some working copy logic and enough things in there that may need to be reproduced.

I guess if the performance is significantly worse we could have a look at this approach.